### PR TITLE
[IMP][12.0] auth_oidc: allow assign groups from token claims

### PR DESCRIPTION
**Pull request by hbrunn**
this is the oauth equivalent of [users_ldap_groups](https://github.com/OCA/server-auth/tree/13.0/users_ldap_groups), allowing to also configure all authorizations in your identity provider.

### CODE REVIEW

File: auth_oidc/views/auth_oauth_provider.xml [hunk 1]
```diff
@@ -19,6 +19,16 @@
                 <field name="token_endpoint" />
                 <field name="jwks_uri" />
             </field>
+            <xpath expr="//sheet/group[last()]" position="after">
+                <group name="mappings">
+                    <field name="group_line_ids">
+                        <tree editable="bottom">
+                            <field name="group_id" />
+                            <field name="expression" />
```
Comment (sbidoul) [inline auth_oidc/views/auth_oauth_provider.xml]: Some help / examples about what to fill-in in `expression` would be welcome.

File: auth_oidc/models/auth_oauth_provider.py [hunk 2]
```diff
@@ -80,3 +84,33 @@ def _parse_id_token(self, id_token, access_token):
...
+                this._eval_expression(self.env.user, {})
+            except (AttributeError, KeyError, NameError) as e:
+                raise exceptions.ValidationError('\n'.join(e.args))
+
+    def _eval_expression(self, user, token):
+        self.ensure_one()
+
+        class Defaultdict2(collections.defaultdict):
+            def __init__(self, *args, **kwargs):
+                super().__init__(Defaultdict2, *args, **kwargs)
```
Comment (sbidoul) [inline auth_oidc/models/auth_oauth_provider.py:110]: Why this Defaultdict2 ?

**Review by sbidoul — COMMENTED** [review]
LGTM. Code review only.

File: auth_oidc/models/auth_oauth_provider.py (hunk 2 above)
Comment (hbrunn) [reply auth_oidc/models/auth_oauth_provider.py:110]: it yields defaultdicts again that in turn yield defaultdicts. This way the user can write expressions in a quite carefree manner like `token['key1']['key2']['key3'] == 'something'` and it doesn't fail when `key1` doesn't exist

Comment (hbrunn) [reply auth_oidc/models/auth_oauth_provider.py:110]: did I do something stupid here?

File: auth_oidc/views/auth_oauth_provider.xml (hunk 1 above)
Comment (hbrunn) [reply auth_oidc/views/auth_oauth_provider.xml]: would

```suggestion
                            <field name="expression" placeholder="token['scope']['value'] == 'hello world'" />
```

suffice?

File: auth_oidc/models/auth_oauth_provider.py (hunk 2 above)
Comment (sbidoul) [reply auth_oidc/models/auth_oauth_provider.py:110]: Nope, just too clever for me :)

File: auth_oidc/views/auth_oauth_provider.xml (hunk 1 above)
Comment (sbidoul) [reply auth_oidc/views/auth_oauth_provider.xml]: Good enough for me :)

File: auth_oidc/models/auth_oauth_provider.py (hunk 2 above)
Comment (hbrunn) [reply auth_oidc/models/auth_oauth_provider.py:110]: just added a comment there

File: auth_oidc/views/auth_oauth_provider.xml (hunk 1 above)
Comment (hbrunn) [reply auth_oidc/views/auth_oauth_provider.xml]: and took this one

**Review by ap-wtioit — APPROVED** [review]
LGTM

**Review by sbidoul — APPROVED** [review]
I had lost track of this. SGTM.
It'd be nice to have a little word about the feature in the README, but not blocking as the view is explicit enough.

**Pull request merged**